src/app-layer-dns-tcp.c

```diff
@@ -47,234 +47,99 @@
-
-    uint16_t q;
-    const uint8_t *data = input + sizeof(DNSHeader);
-
-    //PrintRawDataFp(stdout, (uint8_t*)data, input_len - (data - input));
-
-    if (dns_state != NULL) {
-        if (timercmp(&dns_state->last_req, &dns_state->last_resp, >=)) {
-            if (dns_state->window <= dns_state->unreplied_cnt) {
-                dns_state->window++;
-            }
+    uint64_t tx_id = rs_dns_state_parse_request(dns_state->rs_state, input,
+        input_len);
+
+    /* Drain events. */
+    for (;;) {
+        uint32_t ev = rs_dns_state_get_next_event(dns_state->rs_state);
+        if (ev == 0) {
+            break;
+        } else {
+            DNSSetEvent(dns_state, ev);
         }
     }
 
-    for (q = 0; q < ntohs(dns_header->questions); q++) {
-        uint8_t fqdn[DNS_MAX_SIZE];
-        uint16_t fqdn_offset = 0;
-
-        if (input + input_len < data + 1) {
-            SCLogDebug("input buffer too small for DNSTcpQuery");
-            goto insufficient_data;
-        }
-        SCLogDebug("query length %u", *data);
-
-        while (*data != 0) {
-            if (*data > 63) {
-                /** \todo set event?*/
-                goto insufficient_data;
-            }
-            uint8_t length = *data;
-
-            data++;
-
-            if (length > 0) {
-                if (input + input_len < data + length) {
-                    SCLogDebug("input buffer too small for domain of len %u", length);
-                    goto insufficient_data;
-                }
-                //PrintRawDataFp(stdout, data, qry->length);
-
-                if ((size_t)(fqdn_offset + length + 1) < sizeof(fqdn)) {
-                    memcpy(fqdn + fqdn_offset, data, length);
-                    fqdn_offset += length;
-                    fqdn[fqdn_offset++] = '.';
-                } else {
-                    /** \todo set event? */
-                    goto insufficient_data;
-                }
-            }
-
-            data += length;
-
-            if (input + input_len < data + 1) {
-                SCLogDebug("input buffer too small for DNSTcpQuery(2)");
-                goto insufficient_data;
-            }
-
-            SCLogDebug("qry length %u", *data);
-        }
-        if (fqdn_offset) {
-            fqdn_offset--;
-        }
-
-        data++;
-        if (input + input_len < data + sizeof(DNSQueryTrailer)) {
-            SCLogDebug("input buffer too small for DNSQueryTrailer");
-            goto insufficient_data;
-        }
-        DNSQueryTrailer *trailer = (DNSQueryTrailer *)data;
-        SCLogDebug("trailer type %04x class %04x", ntohs(trailer->type), ntohs(trailer->class));
-        data += sizeof(DNSQueryTrailer);
-
-        /* store our data */
-        if (dns_state != NULL) {
-            DNSStoreQueryInState(dns_state, fqdn, fqdn_offset,
-                    ntohs(trailer->type), ntohs(trailer->class),
-                    ntohs(dns_header->tx_id));
-        }
+    if (tx_id > 0) {
```

Comment:
so I guess the logic here is that after rust ran we need to try to find out if it created a tx and then create our own C owned tx, look up the rust tx and reference it in the C-tx?